### PR TITLE
Add hints to API via the Request interface. Fixes gh-647.

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
Fixes gh-647

### CODE REVIEW

**Comment by OlgaMaciaszek** [comment]
Incomplete. Will reopen after adding more changes.

**Comment by spencergibb** [comment]
I was thinking of having some generics, so you could pass the http request to the loadbalancer for example.

```
public interface Request<C> {

   C getContext();

}
```

I don't want to limit ourselves to a string like ribbon did.

/cc @twicksell

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/DefaultRequest.java [hunk 1]
```diff
@@ -17,8 +17,30 @@
 package org.springframework.cloud.client.loadbalancer.reactive;
 
 /**
+ * A default implementation of {@link Request}.
+ *
  * @author Spencer Gibb
+ * @author Olga Maciaszek-Sharma
  */
-public class DefaultRequest implements Request {
+public class DefaultRequest implements Request<DefaultRequestContext> {
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/DefaultRequest.java]: I think this should be `<T>` still. And then when instantiated user can use `<DefaultRequestContext>`. Maybe?

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/Request.java [hunk 2]
```diff
@@ -20,8 +20,23 @@
...
+	default C getContext() {
+		return null;
+	}
+
+	/**
+	 * A hint that can be used by {@link ReactiveLoadBalancer} implementations to pick the
+	 * correct instance.
+	 * @return a {@link String} representation of the hint
+	 */
+	default String getHint() {
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/Request.java]: Why have two methods?

Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/Request.java]: yes - that's a mistake.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/DefaultRequest.java (hunk 1 above)
Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/DefaultRequest.java]: ok

**Pull request merged**